### PR TITLE
[clang][cas/remote] Introduce a mechanism for storing & retrieving compilation artifacts using an external process service

### PR DESCRIPTION
**Pull request by akyrtzi**
Provide a mechanism for clang to connect to an external process that can store and retrieve compilation artifacts, associated with a particular compilation key. How the external process caches the artifacts is left as implementation detail but the primary motivation is to enable connecting with custom processes that can store and retrieve compilation artifacts via a network distributed cache.

The pieces are:

* `gRPC` service protocol definitions used as the communication API with the external process:
  * API for a key-value service
  * API for a CAS-like service
* a `llvm-remote-cache-test` tool that implements a server for the remote caching protocol, using an on-disk `cas::ObjectStore` and `cas::ActionCache`, intended for testing purposes.
* a remote cache client API for connecting to the external process
* in `cc1_main`, an implementation of `CachingOutputs` that uses the client API for storing and retrieving compilation outputs from the external service

The external service is only used for caching the compilation outputs; dependency scanning, compilation key computation, and the input to the caching compilation is still processed using the `cas::ObjectStore`. Using the local on-disk `cas::ObjectStore` for the inputs ensures efficient processing for the clang caching implementation.

The remote cache mechanism requires `gRPC` to available, and that the CMake variable `LLVM_CAS_ENABLE_REMOTE_CACHE` is enabled.

It is enabled when using the `clang-cache` launcher by setting `LLVM_CACHE_REMOTE_SERVICE_SOCKET_PATH` environment variable to the unix domain socket path that the external service is listening on.

### CODE REVIEW

**Comment by akyrtzi** [comment]
@swift-ci Please test

**Comment by akyrtzi** [comment]
@swift-ci Please test

File: clang/tools/driver/cc1_main.cpp
Comment (cachemeifyoucan) [inline clang/tools/driver/cc1_main.cpp]: This diagnosis and the one below should be emitted in Frontend when validating options.

File: clang/tools/driver/cc1_main.cpp
Comment (cachemeifyoucan) [inline clang/tools/driver/cc1_main.cpp]: What does it mean by having a resultless hit? And why is this a hit as you can't retrieve the output?

File: llvm/cmake/modules/FindGRPC.cmake
Comment (cachemeifyoucan) [inline llvm/cmake/modules/FindGRPC.cmake:90]: What is GPR? Why do we need it?

File: llvm/lib/CAS/ActionCache.cpp
Comment (cachemeifyoucan) [inline llvm/lib/CAS/ActionCache.cpp]: This looks very wrong. Why do we need a CacheKey for any char array? And you can't hash the data before converting to CacheKey, the hashing needs to happen inside the implementation so you won't have two different cache key colliding.

**Review by cachemeifyoucan — COMMENTED** [review]
Looks fine in general.

File: llvm/tools/llvm-remote-cache-test/LLVMCASCacheProvider.cpp
Comment (cachemeifyoucan) [inline llvm/tools/llvm-remote-cache-test/LLVMCASCacheProvider.cpp:106]: Since this is only a test tool, I might just overlay on top of in memory cas since we don't plan to restart the test server and expect data persistency.

File: clang/tools/driver/cc1_main.cpp
Comment (akyrtzi) [reply clang/tools/driver/cc1_main.cpp]: I realize the name `remark_compile_job_cache_hit_resultless` is confusing..

The `remark_compile_job_cache_hit` diagnostic expects to print out the CASID association of "'\<key-casid>' => '\<result-casid>'", but in the case of the remote key-value request there is no "result-casid" to report, so I added `remark_compile_job_cache_hit_resultless` that only prints the "key-casid" (it is a remark diagnostic "without printing the result" / "resultless").

How about to rename to "remark_compile_job_cache_hit_only_key" ?

File: llvm/cmake/modules/FindGRPC.cmake
Comment (akyrtzi) [reply llvm/cmake/modules/FindGRPC.cmake:90]: This is `libgpr.dylib`, part of the set of `gRPC` libraries; I get linker errors for missing symbols `_gpr_log` and `_gpr_unreachable_code`  when linking `llvm-remote-cache-test` without this change. I intend to push this change to upstream as well.

File: llvm/lib/CAS/ActionCache.cpp
Comment (akyrtzi) [reply llvm/lib/CAS/ActionCache.cpp]: > Why do we need a CacheKey for any char array?

This is to be able to implement the remote key-value API using a `cas::ActionCache`, without having to store each individual key in a `cas::ObjectStore`, just to be able to get a `CASID` to use for key.

> And you can't hash the data before converting to CacheKey, the hashing needs to happen inside the implementation so you won't have two different cache key colliding.

It's not clear to me what you mean, could you clarify? With "cache key colliding" do you mean getting the same hash for different set of data? That would mean that we broke the BLAKE3 algorithm and we have bigger problems.

File: llvm/tools/llvm-remote-cache-test/LLVMCASCacheProvider.cpp
Comment (akyrtzi) [reply llvm/tools/llvm-remote-cache-test/LLVMCASCacheProvider.cpp:106]: Actually the lit test (`test/CAS/remote-cache-service.c`) expects being able to connect to a data-persistent service.

File: llvm/lib/CAS/ActionCache.cpp
Comment (cachemeifyoucan) [reply llvm/lib/CAS/ActionCache.cpp]: The data stored in CacheKey is the raw input key. Let's say you somehow mixed CASID and ArrayRef<uint8_t> data and try to store in ActionCache as CacheKey.

CASID is a BLAKE3 hashed value and it is probably computed from another ArrayRef<> Buffer. Now if you store Buffer into the same action cache, Buffer will gets BLAKE3 hashed and it shared the same entry as the CASID computed from buffer.

Comment (akyrtzi) [reply llvm/lib/CAS/ActionCache.cpp]: To make sure I understand, the concern is that:

* I have some ArrayRef data `data1`, that I store in an `ObjectStore` and get a `CASID` that I use as the `CacheKey` for the `ActionCache`; let's call it `key1`.
* I also have some ArrayRef data `data2` that I pass to `CacheKey` directly and that results in `key2`.
* Due to using BLAKE3 for both buffers, if `data1` == `data2` then `key1` == `key2` and we have an undesirable collision of `CacheKey`s.

Did I understand the concern correctly? If yes, then what do you recommend for change?
Should I include some additional static string, like `"__cachekey_from_arrayref__"`, in the hash value of the `ArrayRef<uint8_t>` buffer to avoid this `CacheKey` conflict?

Comment (cachemeifyoucan) [reply llvm/lib/CAS/ActionCache.cpp]: Correct. We need to avoid creating cache key that gets hashed different number of times. It might be fine to do that as implementation details for a test ActionCache but I would not add that to the general API for ActionCache.

Here are some ways to avoid the problem.
* Hash CASID and ArrayRef<uint8_t> one more time before inserting into ActionCache. The original ActionCache interface has this feature so we can probably bring it back if we want to use things other than CASID as key. It would be good to show that we have a real world use case for this, other than the tests.
* Insert ArrayRef into an ObjectStore to get an CASID before insert into ActionCache. You can use a InMemoryCAS, or even implement a nullCAS, which just produce the CASID and threw away the storage.
* For testing, you can just pad your string into desired length and pretend it is a CAS key, or just create a new CASContext for testing (let's say it requires key to be 4 bytes and you can just `CASID::create` them out of thin air) and create a world of ObjectStore/ActionCache out of this world.

I think all of them are acceptable but if you do 1, we need to have a discussion and implement that in a separate PR.

Comment (akyrtzi) [reply llvm/lib/CAS/ActionCache.cpp]: > Due to using BLAKE3 for both buffers, if data1 == data2 then key1 == key2 and we have an undesirable collision of CacheKeys.

Hmm, actually the keys in this case are going to be different, because `BuiltinObjectHasher` also hashes the number of references (in this case `0`), so just hashing the data will result in a different key. Does this change whether there is a concern or not?

> Hash CASID and ArrayRef<uint8_t> one more time before inserting into ActionCache

It's not clear to me why we need to hash twice, wouldn't it be enough to have the `CASContext` contributing a static string for hashing (e.g. "llvmcas"), so that we avoid conflicts between different `CASContext`s? We can treat a standalone buffer as a separate hashing `CASContext` (with a different static string). In any case, we can discuss more in another PR.

> It would be good to show that we have a real world use case for this, other than the tests.

I think this PR does demonstrate a real world use case; I want to implement a generic  key-value API using the `ActionCache` and I'm unable to get a key without storing it into an `ObjectStore`. I should not need to store data for the only reason of passing a key to `ActionCache`.

> Insert ArrayRef into an ObjectStore to get an CASID before insert into ActionCache. You can use a InMemoryCAS, or even implement a nullCAS, which just produce the CASID and threw away the storage.

I can do that for now.

Comment (cachemeifyoucan) [reply llvm/lib/CAS/ActionCache.cpp]: > Does this change whether there is a concern or not?

No. Let me rephrase that. The thing I am against is adding the interface to CacheKey class that introduces the wrong behavior, which is the raw bytes stored in `CacheKey` can mean different things and can go through different amount of hash functions. `CacheKey::CacheKey(ArrayRef<uint8_t> Data)` is fine, but constructing using BLAKE3 is not.

I think a new CASContext for your wrapper ActionCache is the fitting concept to use. But I forgot that we mentioned that we should maybe put two CASContext into ActionCache to represent different contexts of keys and values. Let me know if that is too hard so we can adjust the ActionCache correspondingly.

Comment (akyrtzi) [reply llvm/lib/CAS/ActionCache.cpp]: I mainly want the `CacheKey(ArrayRef<uint8_t> Data)` constructor to be available for clients. If that requires a `CASContext` it should be an implementation detail of the `ActionCache`.

For the purposes of this PR I'm going to be using an in-memory `ObjectStore` in order to get a key.

Comment (akyrtzi) [reply llvm/lib/CAS/ActionCache.cpp]: See new commit: https://github.com/apple/llvm-project/pull/5355/commits/03fa5cde0aeaacda35703f0e002a85bcae00434b

Comment (akyrtzi) [reply llvm/lib/CAS/ActionCache.cpp]: > I mainly want the CacheKey(ArrayRef<uint8_t> Data) constructor to be available for clients.

I would also be happy with `ActionCache` methods that I can use to form keys, it's not strictly necessary that the `CacheKey` must be formed using context-free constructors.

File: clang/tools/driver/cc1_main.cpp
Comment (akyrtzi) [reply clang/tools/driver/cc1_main.cpp]: Good idea! See https://github.com/apple/llvm-project/pull/5355/commits/a5b724b9b53c2d2c84d5708237ce137f1d40f5ca

File: clang/tools/driver/cc1_main.cpp
Comment (cachemeifyoucan) [reply clang/tools/driver/cc1_main.cpp]: I don't quite understand why there is no "result-casid"? Because the result id can only be decoded in replay function? Sounds like we should get that ID out of replay function.

Comment (akyrtzi) [reply clang/tools/driver/cc1_main.cpp]: In the remote key-value API the value that you get is a dictionary (`message Value { map<string, bytes> entries = 1; }`), there is no single CASID to report.

Comment (cachemeifyoucan) [reply clang/tools/driver/cc1_main.cpp]: I know about that, I am just wondering if we should surface out the CASID for output in the dictionary.

Comment (akyrtzi) [reply clang/tools/driver/cc1_main.cpp]: You mean dump out the whole dictionary? I don't think it's that useful, in practice I haven't encountered an actual need to use the printed result CASID. But we can re-evaluate if we find a compelling use, probably with a separate remark for enabling even more detailed/verbose diagnostic output.

Comment (cachemeifyoucan) [reply clang/tools/driver/cc1_main.cpp]: I mean just the result CASID for the output part. I know it is useful for me when debugging nondeterminism because if you have access to the CAS, you can dump out the two different result CASID and inspect for difference.

Comment (akyrtzi) [reply clang/tools/driver/cc1_main.cpp]: Do you mean just print the `bytes` for the entry with `"<output>"` key, ignoring the rest?

Comment (cachemeifyoucan) [reply clang/tools/driver/cc1_main.cpp]: Yup, maybe `replayCachedResult` can return `Expected<Optional<CASID>>`?

**Review by cachemeifyoucan — APPROVED** [review]
Other than the current 1 comment out diagnostics message without CASID, all other changes LGTM.

File: clang/test/CAS/remote-cache-incompatible-options.c
Comment (benlangmuir) [inline clang/test/CAS/remote-cache-incompatible-options.c]: nit: Extra newline

File: clang/test/CAS/remote-cache-service.c
Comment (benlangmuir) [inline clang/test/CAS/remote-cache-service.c:3]: We should set this up in in `test/CAS/lit.local.cfg` like we do for `clang-cc1daemon`.

File: clang/tools/driver/cc1_main.cpp
Comment (benlangmuir) [inline clang/tools/driver/cc1_main.cpp:293]: How hard would it be to split all the caching backend code out of cc1_main.cpp? Ideally we could have one .cpp for each caching backend plus one for the common code. Upstream this file is only ~300 lines, but here it is now ~1200.  It seems bad to have cc1_main be 75% caching-related.

Comment (akyrtzi) [reply clang/tools/driver/cc1_main.cpp:293]: Good idea, this seems straightforward by pulling the classes out of `cc1_main.cpp`.
But I'd like to do that as follow-up after eliminating any internal merge conflict.

Comment (benlangmuir) [reply clang/tools/driver/cc1_main.cpp:293]: sgtm

File: clang/tools/driver/cc1_main.cpp
Comment (akyrtzi) [reply clang/tools/driver/cc1_main.cpp]: Sounds good! See https://github.com/apple/llvm-project/pull/5355/commits/67709c67630a9cf21083e611b594554695c80cdb
I moved printing the remark in `replayCachedResult()`, which also helps for printing the remark before the cached stderr.

File: clang/test/CAS/remote-cache-incompatible-options.c
Comment (akyrtzi) [reply clang/test/CAS/remote-cache-incompatible-options.c]: Fixed.

File: clang/test/CAS/remote-cache-service.c
Comment (akyrtzi) [reply clang/test/CAS/remote-cache-service.c:3]: Thanks for the pointer! See https://github.com/apple/llvm-project/pull/5355/commits/6128e9f2ae1e573e70bb6bb224e131ce5d08a485

**Comment by benlangmuir** [comment]
The issues I mentioned are fixed.  I haven't reviewed the rest of this code, but don't block on me.

**Comment by akyrtzi** [comment]
@swift-ci Please test

**Comment by akyrtzi** [comment]
@swift-ci Please test

**Pull request merged**
